src/com/amazon/corretto/crypto/provider/AesKeyWrapSpi.java

```diff
@@ -166,6 +195,86 @@ private void implInit(int opmode, Key key, SecureRandom ignored)
             jceKey = (SecretKey) key;
         }
         this.opmode = opmode;
+        buffer.reset();
+    }
+
+    @Override
+    protected byte[] engineUpdate(byte[] in, int inOfs, int inLen) {
+        if (opmode != Cipher.ENCRYPT_MODE && opmode != Cipher.DECRYPT_MODE) {
+            throw new IllegalStateException("Cipher not initialized for update");
+        }
+        implUpdate(in, inOfs, inLen);
+        return null;
+    }
+
+    @Override
+    protected int engineUpdate(byte[] in, int inOfs, int inLen, byte[] out, int outOf)
+            throws ShortBufferException {
+        if (opmode != Cipher.ENCRYPT_MODE && opmode != Cipher.DECRYPT_MODE) {
+            throw new IllegalStateException("Cipher not initialized for update");
+        }
+        // TODO [childw] check for short output buffer
+        implUpdate(in, inOfs, inLen);
+        return 0;
+    }
+
+    private void implUpdate(byte[] in, int inOfs, int inLen) {
+        buffer.write(in, inOfs, inLen);
+    }
+
+    @Override
+    protected byte[] engineDoFinal(byte[] in, int inOfs, int inLen) {
+        if (opmode != Cipher.ENCRYPT_MODE && opmode != Cipher.DECRYPT_MODE) {
+            throw new IllegalStateException("Cipher not initialized for finalization");
+        }
+        return implDoFinal(in, inOfs, inLen);
+    }
+
+    @Override
+    protected int engineDoFinal(byte[] in, int inOfs, int inLen, byte[] out, int outOfs) {
+        if (opmode != Cipher.ENCRYPT_MODE && opmode != Cipher.DECRYPT_MODE) {
+            throw new IllegalStateException("Cipher not initialized for finalization");
+        }
+        return implDoFinal(in, inOfs, inLen, out, outOfs);
+    }
+
+    private byte[] implDoFinal(byte[] in, int inOfs, int inLen) {
+        final int estimatedOutLen = engineGetOutputSize(inLen - inOfs);
+        byte[] out = new byte[estimatedOutLen];
+        final int actualOutLen = implDoFinal(in, inOfs, inLen, out, 0);
+        // If we overestimated the size of the output (possible in unwrapping),
+        // we need to copy only the output's bytes over to a newer, smaller
+        // byte array. Java's inability to truncate arrays after creation
+        // forces us to do this. Note that in the common case of block-aligned
+        // key sizes, our estimates are correct and this extra copy is avoided.
+        if (actualOutLen < estimatedOutLen) {
+            final byte[] tmp = new byte[actualOutLen];
+            System.arraycopy(out, 0, tmp, 0, tmp.length);
+            Arrays.fill(out, (byte) 0);
+            out = tmp;
+        }
+        return out;
+    }
+
+    private int implDoFinal(byte[] in, int inOfs, int inLen, byte[] out, int outOfs) {
+        implUpdate(in, inOfs, inLen);
+        buffer.trim();
```

Comment:
Why do this instead of passing down the length as a parameter? Trimming costs extra byte copies and you can pass in a maximum length when instantiating `java_buffer`.

Reply:
ah, nice. i didn't see that overload of `java_buffer::from_array`. thanks for the tip.